File: src/main/c/jni-lib/src/events/OptionSaleMapping.cpp
// Copyright (c) 2024 Devexperts LLC.
// SPDX-License-Identifier: MPL-2.0

#include "dxfeed/events/OptionSaleMapping.hpp"
#include "dxfeed/utils/ByteReader.hpp"
#include "dxfeed/utils/ByteWriter.hpp"

namespace dxfeed::jni {
dxfg_option_sale_t *OptionSaleMapping::toOptionSale(ByteReader &reader) {
    auto *optionSale = new dxfg_option_sale_t();
    optionSale->market_event.event_type.clazz = DXFG_EVENT_OPTION_SALE;
    optionSale->market_event.event_symbol = reader.readString();
    optionSale->market_event.event_time = reader.readLong();
    optionSale->event_flags = reader.readInt();
    optionSale->index = reader.readLong();
    optionSale->time_sequence = reader.readLong();
    optionSale->time_nano_part = reader.readInt();
    optionSale->exchange_code = reader.readInt16_t();
    optionSale->flags = reader.readInt();
    optionSale->exchange_sale_conditions = reader.readString();
    optionSale->option_symbol = reader.readString();

    optionSale->price = reader.readDouble();
    optionSale->size = reader.readDouble();
    optionSale->bid_price = reader.readDouble();
    optionSale->ask_price = reader.readDouble();
    optionSale->underlying_price = reader.readDouble();
    optionSale->volatility = reader.readDouble();
    optionSale->delta = reader.readDouble();
    return optionSale;
}

void OptionSaleMapping::fromOptionSale(dxfg_option_sale_t *optionSale, ByteWriter &writer) {
    writer.writeString(optionSale->market_event.event_symbol);
    writer.writeInt64_t(optionSale->market_event.event_time);
    writer.writeInt32_t(optionSale->event_flags);
    writer.writeInt64_t(optionSale->index);
    writer.writeInt64_t(optionSale->time_sequence);
    writer.writeInt32_t(optionSale->time_nano_part);
    writer.writeInt16_t(optionSale->exchange_code);
    writer.writeInt32_t(optionSale->flags);
    writer.writeString(optionSale->exchange_sale_conditions);
    writer.writeString(optionSale->option_symbol);

    writer.writeDouble(optionSale->price);
    writer.writeDouble(optionSale->size);
    writer.writeDouble(optionSale->bid_price);
    writer.writeDouble(optionSale->ask_price);
    writer.writeDouble(optionSale->underlying_price);
    writer.writeDouble(optionSale->volatility);
    writer.writeDouble(optionSale->delta);
}
} // namespace dxfeed::jni

File: src/main/java/com/dxfeed/api/serializers/OptionSaleMapping.java
package com.dxfeed.api.serializers;

import com.dxfeed.api.buffers.ByteBuffer;
import com.dxfeed.api.buffers.DoubleBuffer;
import com.dxfeed.api.buffers.ByteReader;
import com.dxfeed.event.market.DxFeedEventMarketPackagePrivate;
import com.dxfeed.event.market.OptionSale;

public class OptionSaleMapping {
  /**
   * typedef struct dxfg_market_event_t {
   *    dxfg_event_type_t event_type;
   *    const char *event_symbol;
   *    int64_t event_time;
   * } dxfg_market_event_t;
   *
   * https://github.com/dxFeed/dxfeed-graal-native-sdk/blob/main/src/main/c/api/dxfg_events.h#L360
   *
   * typedef struct dxfg_option_sale_t {
   *    dxfg_market_event_t market_event;
   *    //    dxfg_indexed_event_t indexed_event;
   *    int32_t event_flags;
   *    int64_t index;
   *    int64_t time_sequence;
   *    int32_t time_nano_part;
   *    int16_t exchange_code;
   *    double price;
   *    double size;
   *    double bid_price;
   *    double ask_price;
   *    const char *exchange_sale_conditions;
   *    int32_t flags;
   *    double underlying_price;
   *    double volatility;
   *    double delta;
   *    const char *option_symbol;
   * } dxfg_option_sale_t;
   */

  public static void toNative(OptionSale event, ByteBuffer pBytes, DoubleBuffer pDoubles) {
    // BYTE DATA
    pBytes.writeString(event.getEventSymbol());                           // 2 + eventSymbolLength
    pBytes.writeLong(event.getEventTime());                               // 8
    pBytes.writeInt(event.getEventFlags());                               // 4
    pBytes.writeLong(event.getIndex());                                   // 8
    pBytes.writeLong(event.getTimeSequence());                            // 8
    pBytes.writeInt(event.getTimeNanoPart());                             // 4
    pBytes.writeChar(event.getExchangeCode());                            // 2
    pBytes.writeInt(DxFeedEventMarketPackagePrivate.getFlags(event));     // 4
    pBytes.writeString(event.getExchangeSaleConditions());                // 2 + exchangeSaleConditionsLength
    pBytes.writeString(event.getOptionSymbol());                          // 2 + optionSymbolLength

    // DOUBLE DATA
    pDoubles.write(event.getPrice());
    pDoubles.write(event.getSize());
    pDoubles.write(event.getBidPrice());
    pDoubles.write(event.getAskPrice());
    pDoubles.write(event.getUnderlyingPrice());
    pDoubles.write(event.getVolatility());
    pDoubles.write(event.getDelta());
  }

  public static OptionSale fromNative(ByteReader reader) {
    OptionSale optionSale = new OptionSale();
    optionSale.setEventSymbol(reader.readString());
    optionSale.setEventTime(reader.readLong());
    optionSale.setEventFlags(reader.readInt());
    optionSale.setIndex(reader.readLong());
    optionSale.setTimeSequence(reader.readLong());
    optionSale.setTimeNanoPart(reader.readInt());
    optionSale.setExchangeCode(reader.readChar());
    DxFeedEventMarketPackagePrivate.setFlags(optionSale, reader.readInt());
    optionSale.setExchangeSaleConditions(reader.readString());
    optionSale.setOptionSymbol(reader.readString());

    optionSale.setPrice(reader.readDouble());
    optionSale.setSize(reader.readDouble());
    optionSale.setBidPrice(reader.readDouble());
    optionSale.setAskPrice(reader.readDouble());
    optionSale.setUnderlyingPrice(reader.readDouble());
    optionSale.setVolatility(reader.readDouble());
    optionSale.setDelta(reader.readDouble());
    return optionSale;
  }
}
